File: 1-working-with-list/p23_rand_select.ml

let rec at n l =
  match l with
  | [] -> None
  | hd :: tail when n > 0 -> if n = 1 then Some hd else at (n-1) tail
  | _ -> None ;;

let remove_at n l =
  let rec aux acc curr = function
  | [] -> List.rev acc
  | h :: t ->
    if curr = 0 then
      (List.rev acc) @ t
    else
      aux (h::acc) (curr-1) t
  in aux [] n l;;

let rand_select lst n =
  let rec aux acc curr = function
    | [] ->
      acc
    | l ->
      if curr = 0 then
        acc
      else
        let i = Random.int (List.length l) in
        let e = match (at (i+1) l) with Some(x) -> x | None -> assert false in
        aux (e::acc) (curr-1) (remove_at i l)
  in aux [] n lst;;

rand_select ["a";"b";"c";"d";"e";"f";"g";"h"] 3;;
rand_select ["a";"b";"c"] 3;;
rand_select ["a";"b"] 3;;

File: 1-working-with-list/p23_rand_select.scala

object Main {

  def main(args: Array[String]) {
    val l = List("a", "b", "c", "d")
    println(extractAt(0, l))
    println(extractAt(1, l))
    println(extractAt(3, l))
    println(extractAt(6, l))

    val m = List("a", "b", "c", "d", "e", "f", "g", "h")
    println(randSelect(m, 3))
  }

  def extractAt[T](index: Int, list: List[T]): (List[T], Option[T]) = {
    def aux(l: List[T], c: Int, acc: List[T]): (List[T], Option[T]) = {
      l match {
        case Nil => (acc.reverse, None)
        case hd :: tail =>
          if(c == 0)
            (acc.reverse ::: tail, Some(hd))
          else
            aux(tail, c-1, hd :: acc)
      }
    }
    aux(list, index, List())
  }

  def randSelect[T](l: List[T], i: Int): List[T] = {
    val r = (new scala.util.Random).nextInt(l.length)
    if(i > 0)
      extractAt(r, l) match {
        case (_, None) => List()
        case (rest, Some(e)) => e :: randSelect(rest, i-1)
      }
    else
      List()
  }
}
